File: Laba2/Laba2/Heatovoz.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Laba2
{
    class Heatovoz : Locomotive, IComparable<Heatovoz>, IEquatable<Heatovoz>
    {
        private bool BotmChimney;

        private bool TopChimney;

        private int countFuel;

        private Color chimneyColor;

        public Heatovoz(int maxSpeed, int maxCountPassengers, double weight, double carrying, Color color, Color dopColor, bool BotmChimney, bool TopChimney, int countFuel, Color chimneyColor) :
            base(maxSpeed, maxCountPassengers, weight, carrying, color, dopColor)
        {
            this.BotmChimney = BotmChimney;
            this.TopChimney = TopChimney;
            this.chimneyColor = chimneyColor;
            this.countFuel = countFuel;
        }

        public Heatovoz(string info) : base(info)
        {
            string[] properties = info.Split(';');
            if (properties.Length == 10)
            {
                MaxSpeed = Convert.ToInt32(properties[0]);
                MaxCountPassengers = Convert.ToInt32(properties[1]);
                Weight = Convert.ToDouble(properties[2]);
                Carring = Convert.ToDouble(properties[3]);
                ColorBody = Color.FromName(properties[4]);
                DopColor = Color.FromName(properties[5]);
                BotmChimney = Convert.ToBoolean(properties[6]);
                TopChimney = Convert.ToBoolean(properties[7]);
                countFuel = Convert.ToInt32(properties[8]);
                chimneyColor = Color.FromName(properties[9]);
            }
        }

        public int CompareTo(Heatovoz other)
        {
            var res = (this is Locomotive).CompareTo(other is Locomotive);
            if (res != 0)
            {
                return res;
            }
            if (BotmChimney != other.BotmChimney)
            {
                return BotmChimney.CompareTo(other.BotmChimney);
            }
            if (TopChimney != other.TopChimney)
            {
                return TopChimney.CompareTo(other.TopChimney);
            }
            if (countFuel != other.countFuel)
            {
                return countFuel.CompareTo(other.countFuel);
            }
            if (chimneyColor != other.chimneyColor)
            {
                return chimneyColor.Name.CompareTo(other.chimneyColor.Name);
            }
            return 0;
        }

        public bool Equals(Heatovoz other)
        {
            var res = (this is Locomotive).Equals(other is Locomotive);
            if (!res)
            {
                return res;
            }
            if (TopChimney != other.TopChimney)
            {
                return false;
            }
            if (BotmChimney != other.BotmChimney)
            {
                return false;
            }
            if(chimneyColor != other.chimneyColor)
            {
                return false;
            }
            if(countFuel != other.countFuel)
            {
                return false;
            }
            return true;
        }

        public override bool Equals(object obj)
        {
            if(obj == null)
            {
                return false;
            }
            Heatovoz locObj = obj as Heatovoz;
            if(locObj == null)
            {
                return false;
            }else
            {
                return Equals(locObj);
            }
        }
        public override int GetHashCode()
        {
            return base.GetHashCode();
        }
        protected override void drawLocomotive(Graphics g)
        {
            base.drawLocomotive(g);

            Brush br;

            if (BotmChimney)
            {
                br = new SolidBrush(chimneyColor);
                Rectangle[] botmChimney = new Rectangle[2];
                botmChimney[0] = new Rectangle((bodyWidth >> 1) - bodyWidth / 3 + 2, (bodyHeight >> 1) - bodyHeight / 12 + 2,
                    2 * bodyWidth / 3 - 4, 2 * bodyWidth / 3 - 4);
                botmChimney[1] = new Rectangle(botmChimney[0].X + 1, botmChimney[0].Y + 1, botmChimney[0].Width - 2, botmChimney[0].Height - 2);
                g.FillEllipse(br, botmChimney[0]);
                br = new SolidBrush(Color.Black);
                g.FillEllipse(br, botmChimney[1]);
            }

            if (TopChimney)
            {
                br = new SolidBrush(chimneyColor);
                Rectangle[] topChimney = new Rectangle[2];
                topChimney[0] = new Rectangle((bodyWidth >> 1) - bodyWidth / 3 + 2, (bodyHeight >> 1) - bodyHeight / 4 + 2,
                    2 * bodyWidth / 3 - 4, 2 * bodyWidth / 3 - 4);
                topChimney[1] = new Rectangle(topChimney[0].X + 1, topChimney[0].Y + 1,
                    topChimney[0].Width - 2, topChimney[0].Height - 2);
                g.FillEllipse(br, topChimney[0]);
                br = new SolidBrush(Color.Black);
                g.FillEllipse(br, topChimney[1]);
            }
        }

        public void setChimneyColor(Color chimneyColor)
        {
            this.chimneyColor = chimneyColor;
        }
        public override string getInfo()
        {
            return base.getInfo() + ";" + BotmChimney + ";" + TopChimney + ";" + countFuel + ";" + chimneyColor.Name;
        }
    }
}

File: Laba2/src/main/Heatovoz.java
package main;

import java.awt.*;

public class Heatovoz extends Locomotive {

    private boolean botmChimney;

    private boolean topChimney;

    private int countFuel;

    private Color chimneyColor;

    public Heatovoz(int maxCountPassengers,
                    int maxSpeed,
                    double carrying,
                    double weight,
                    Color colorBody,
                    Color dopColor,
                    boolean botmChimney,
                    boolean topChimney,
                    int countFuel,
                    Color chimneyColor) {
        super(maxCountPassengers,
                maxSpeed,
                carrying,
                weight,
                colorBody,
                dopColor);
        this.botmChimney=botmChimney;
        this.topChimney=topChimney;
        this.countFuel=countFuel;
        this.chimneyColor=chimneyColor;
    }
    @Override
    protected void drawLocomotive(Graphics g){
        super.drawLocomotive(g);

        if (botmChimney)
        {
            g.setColor(chimneyColor);
            Rectangle[] botmChimney = new Rectangle[2];
            botmChimney[0] = new Rectangle((getBodyWidth()>>1) - getBodyWidth() / 3 + 2, (getBodyHeight()>>1) - getBodyHeight() / 12 + 2,
                    2 * getBodyWidth() / 3 - 4, 2 * getBodyWidth() / 3 - 4);
            botmChimney[1] = new Rectangle(botmChimney[0].x + 1, botmChimney[0].y + 1, botmChimney[0].width - 2, botmChimney[0].height - 2);
            g.fillOval(botmChimney[0].x,botmChimney[0].y,botmChimney[0].width,botmChimney[0].height);
            g.setColor(Color.BLACK);
            g.fillOval(botmChimney[1].x,botmChimney[1].y,botmChimney[1].width,botmChimney[1].height);
        }

        if (topChimney)
        {
            g.setColor(chimneyColor);
            Rectangle[] topChimney = new Rectangle[2];
            topChimney[0] = new Rectangle((getBodyWidth()>>1) - getBodyWidth() / 3 + 2, (getBodyHeight()>>1) - getBodyHeight() / 4 + 2,
                    2 * getBodyWidth() / 3 - 4, 2 * getBodyWidth() / 3 - 4);
            topChimney[1] = new Rectangle(topChimney[0].x + 1, topChimney[0].y + 1,
                    topChimney[0].width - 2, topChimney[0].height - 2);
            g.fillOval(topChimney[0].x,topChimney[0].y,topChimney[0].width,topChimney[0].height);
            g.setColor(Color.black);
            g.fillOval(topChimney[1].x,topChimney[1].y,topChimney[1].width,topChimney[1].height);
        }

    }
}
